Sources/GRPCCore/GRPCClient.swift

```diff
@@ -112,34 +112,53 @@ public final class GRPCClient: Sendable {
   /// The transport which provides a bidirectional communication channel with the server.
   private let transport: any ClientTransport
 
-  /// A collection of interceptors providing cross-cutting functionality to each accepted RPC.
-  ///
-  /// The order in which interceptors are added reflects the order in which they are called. The
-  /// first interceptor added will be the first interceptor to intercept each request. The last
-  /// interceptor added will be the final interceptor to intercept each request before calling
-  /// the appropriate handler.
-  private let interceptors: [any ClientInterceptor]
+  private let interceptorPipeline: [ClientInterceptorPipelineOperation]
 
   /// The current state of the client.
   private let state: Mutex<State>
 
   /// The state of the client.
   private enum State: Sendable {
+    
     /// The client hasn't been started yet. Can transition to `running` or `stopped`.
-    case notStarted
+    case notStarted(
+      /// A collection of interceptors providing cross-cutting functionality to each accepted RPC, keyed by the method to which they apply.
+      ///
+      /// The list of interceptors for each method is computed from `interceptorsPipeline` when calling a method for the first time.
+      /// This caching is done to avoid having to compute the applicable interceptors for each request made.
+      ///
+      /// The order in which interceptors are added reflects the order in which they are called. The
+      /// first interceptor added will be the first interceptor to intercept each request. The last
+      /// interceptor added will be the final interceptor to intercept each request before calling
+      /// the appropriate handler.
+      interceptorsPerMethod: [MethodDescriptor: [any ClientInterceptor]]
```

Comment:
The current state is orthogonal to the cache, rather than storing the cache in the enum cases, can we wrap `State` in a `StateMachine` struct and store the cache in the `StateMachine` struct as a sibling to the `State`? It also means we don't need the `_modifying` state.

```swift
struct StateMachine {
  var state: State
  var interceptors:  [MethodDescriptor: [any ClientInterceptor]]
}
```